readthedocs/projects/tasks/utils.py

```diff
@@ -154,6 +160,87 @@ def send_external_build_status(version_type, build_pk, commit, status):
         send_build_status.delay(build_pk, commit, status)
 
 
+class DeprecatedConfigFileSiteNotification(SiteNotification):
+
+    failure_message = (
+        "Your project '{{ object.slug }}' doesn't have a "
+        '<a href="https://docs.readthedocs.io/en/stable/config-file/v2.html">.readthedocs.yaml</a> '
+        "configuration file. "
+        "Configuration files will <strong>soon be required</strong> by projects, "
+        "and will no longer be optional. "
+        "Make sure to create one for your project to ensure your project continues "
+        "building successfully."
+    )
+    failure_level = WARNING_PERSISTENT
+
+
+class DeprecatedConfigFileEmailNotification(Notification):
+
+    app_templates = "projects"
+    name = "deprecated_config_file_used"
+    context_object_name = "project"
+    subject = "Your project will start failing soon"
+    level = REQUIREMENT
+
+    def send(self):
+        """Method overwritten to remove on-site backend."""
+        backend = EmailBackend(self.request)
+        backend.send(self)
+
+
+@app.task(queue="web")
+def deprecated_config_file_used_notification():
+    """
+    Create a notification about not using a config file for all the maintainers of the project.
+
+    This is a scheduled task to be executed on the webs.
+    Note the code uses `.iterator` and `.only` to avoid killing the db with this query.
+    """
+    projects = set()
+    # NOTE: we could skip the projects with a spam score > 150,
+    # to reduce the amount of email/onsite notifications we send
+    start_datetime = datetime.datetime.now()
+    for project in Project.objects.all().only("slug", "default_version").iterator():
+        # NOTE: instead of iterating over all the active versions,
+        # we can only consider the default one
+        for version in (
+            project.versions.filter(Q(active=True) | Q(slug=project.default_version))
+            # Add a limit of 15 to protect ourselves against projects with
+            # hundred of active versions
+            .only("id")[:15].iterator()
+        ):
+            build = version.builds.filter(success=True).only("_config").first()
+            if build and build.deprecated_config_used():
+                projects.add(project.slug)
+                # Do not continue iterating over the
+                # other versions when we know this project
+                # will get the notification
+                break
+
+    log.info(
+        "Sending deprecated config file notification.",
+        query_seconds=(datetime.datetime.now() - start_datetime).seconds,
+        projects=len(projects),
+    )
+
+    for project in Project.objects.filter(slug__in=projects):
+        users = AdminPermission.owners(project)
+        for user in users:
+            n = DeprecatedConfigFileSiteNotification(
+                user=user,
+                context_object=project,
+                success=False,
+            )
+            n.send()
+
+        for user in users:
```

Comment:
Is there performance gain to only iterating over `users` once? Seems these loops could be joined.